### PR TITLE
TCP UI + Interception

### PR DESCRIPTION
**Pull request by Jessonsotoventura**
This PR is a WIP.

It adds a new entry into the Flowlist to render TCP connections (currently limited to table view). It also adds the ability to intercept and modify TCP connections.

Usages:
mitmproxy --mode transparent --rawtcp --tcp-hosts '<hosts>'

Notes:
TCP and HTTP requests render in the same interface view.
There are still a ton of parts where TCP crashes, but currently the following all work:
- TCP interception
- TCP UI (Only tabular view has been worked on thus far)
- TCP Detailed View with some view modes (raw, hex, maybe others)
- TCP modification
- Some, but not all filters

Any feedback and comments on the features would be helpful.

### CODE REVIEW

**Comment by Kriechi** [comment]
Hi @Jessonsotoventura!
Thanks for sending this PR - @mhils and I will try to find some time to review it properly.
By any chance, did you take a look at https://github.com/mitmproxy/mitmproxy/pull/3626 - there are similar refactorings and addiions - maybe you could join forces with @Josue-T ?

File: mitmproxy/flowfilter.py [hunk 1]
```diff
@@ -76,6 +76,13 @@ class _Action(_Token):
     def make(klass, s, loc, toks):
         return klass(*toks[1:])
 
+class FAll(_Action):
+    code = "all"
+    help = "Show all types of flows"
+
+    @only(tcp.TCPFlow, http.HTTPFlow, websocket.WebSocketFlow)
+    def __call__(self, f):
+        return True
```
[the remark this answers is not in the document]
Comment (Jessonsotoventura) [reply mitmproxy/flowfilter.py]: Yes - The current naked filter is FURL. This means that TCP traffic is not displayed unless the ~tcp filter is applied. I think that if there is a single view then it should show all traffic by default. Thus the FALL to show all types of traffic.

[the remark this answers is not in the document]
Comment (mhils) [reply mitmproxy/flowfilter.py]: Ah, totally missed that change below. Makes sense 👍 
This change makes it eat up naked filters, which is a bit annoying. We can probably introduce something like FDefault, which matches on URLs for HTTPFlows and on message content for TCPFlows. Don't worry about it if you're not into the flowfilter part, I can take care of it when we merge. :)

File: mitmproxy/tcp.py [hunk 2]
```diff
@@ -1,22 +1,40 @@
...
 
     @classmethod
     def from_state(cls, state):
         return cls(*state)
 
+    @property
+    def content(self):
+        if self.encoding is None:
+            return  self._raw_content
+        #TODO ADD ENCODING SUPPORT
```
[the remark this answers is not in the document]
Comment (Jessonsotoventura) [reply mitmproxy/tcp.py]: Yeah - I'm not entire certain about this either just yet. Probably will shelf it for now

File: mitmproxy/tcp.py [hunk 3]
```diff
@@ -36,12 +54,190 @@ class TCPFlow(flow.Flow):
     A TCPFlow is a simplified representation of a TCP session.
     """
 
-    def __init__(self, client_conn, server_conn, live=None):
+    def __init__(self, client_conn, server_conn, live=None, index=0):
         super().__init__("tcp", client_conn, server_conn, live)
-        self.messages: List[TCPMessage] = []
+        self._messages =  dict() 
+        self._message_order = list()
```
[the remark this answers is not in the document]
Comment (Jessonsotoventura) [reply mitmproxy/tcp.py]: Yeah - I've removed this already in my local branch.

**Pull request closed without merging**